IdentityCore/src/cache/ios/MSIDKeychainTokenCache.m

```diff
@@ -744,5 +745,18 @@ - (MSIDCacheKey *)overrideTokenKey:(MSIDCacheKey *)key
     return key;
 }
 
+- (NSString *)extractAppKey:(NSString *)cacheKeyString
+{
+    if ([NSString msidIsStringNilOrBlank:cacheKeyString]) return nil;
+    
+    NSArray *components = [cacheKeyString componentsSeparatedByString:@"|"];
+    
+    // app key is added as the 2nd part (default tokens) and 5th part (legacy tokens) by keychain token cache
+    if ((components.count == 2 || components.count == 5) && ![NSString msidIsStringNilOrBlank:components.lastObject])
```

Comment:
What are the benefits of checking exact component count vs taking the last element? Will ADAL always have 5 elements?

Reply:
It will be problematic if scopes contain "|".

I am now still thinking about the following two solutions:
1) Store app key to the cache item. Then the app key extraction logic is not needed by default tokens anymore but only needed by legacy token, and the logic could be moved to legacy serializer.

2) Move the key extraction logic to `ADBrokerKeychainTokenCache`, such that we could always return the last element as the app key.

What do you think?

Reply:
@jasoncoolmax, let's discuss offline.

Reply:
discussed offline, we took the second approach, which makes more sense :)